[Review Ready]Wing mutant body parts and Angel wings

Shadowlight213 opened the pull request:
Currently I only have the basic wing mutant body parts done.

Angel wings are going to be used for a future Archangel antag
The archangel is sort of going to be the divine judgement against the sintouched.
Sort of an opposite to the lawyer demon.
Angel wings are going to be admin/archangelonly however roundstart wings is code supported in case medbay wants to be ~~succubi~~ saviors.

I'm mainly working on the wing component

Update: I've refactored the Angel flight abilities to be based on a species. This should make things way less hacky.
Progress:
Angel wings themselves will offer a number of advantages and disadvantages:
- [x] Angel wings act like jetpacks while flying in an atmosphere
- [x] Being stunned while flying causes you to lose control and be lube launched and stunned for a long time

~~\- [ ] True demon can completely rip an angel's wings off if they get them stunned.~~ This is going to need more sprites to do, so it will have to be done in a later PR.

Wings general:
- [x] Actually get sprites that aren't the griffon cloak
- [x] Rework mutant body parts to allow for flagging parts as restricted from roundstart

Code cleanup:
- [x] See if it's viable to datumize the flight system.
- [x] Change everything to a species so it's a million times cleaner.

EXPERIMENTAL:
~~\- [ ] See if it's possible to make it an organ.~~
This is going to need more sprites to do, so it will have to be done in a later PR.

Kickass sprites done by Kor's GF.

![wings](http://puu.sh/oMuxp/ce9757ddb8.jpg)

Shadowlight213 committed:
Angel WIP commit 1
Added basic angel wing Mutant parts

Dorsisdwarf commented:
My god, it's what the mutant bodypart system was allegedly for!

KorPhaeron commented:
>however roundstart wings is code supported in case medbay wants to be succubi saviors.

@Iamgoofball was right

KazeEspada commented:
Yay more customization options for the most racist server.

FantasticFwoosh commented:
Pass the redbull over here matey. It's given us wings.

xxalpha commented:
>wings

uhhhhh

Core0verload commented:
Mutant parts really, really need a rewrite, :+1: 

However, eventually this will result in roundstart angel catgirls.

TechnoAlchemisto commented:
t-they'll never be roundstart, right?

Iamgoofball commented:
# SLIPPERY

#SLOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOÒOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOPE

Shadowlight213 commented:
Archangel wings will definitely be admin only. They provide too many advantages to be roundstart.
Cosmetic wings..probably not roundstart.
Remember, open wings are going to drastically increase your hitbox for melee combat. It's a mechanical nerf to play with them.

TechnoAlchemisto commented:
so are cat ears and lizard tails

TrustyGun commented:
we hippiestation now

Jazaen commented:
>Archangel antag
Hans... Are we the baddies?

octareenroon91 commented:
>holy antag

It's a valids/faction issue.

Angels that happen to be on the station do not share the same contractual relationship of trust that binds all crewmembers together.

Centcom roles might also be called 'antag' for that reason.



Shadowlight, do you think you can fit in fly wings?

Jazaen commented:
Not really. "Antag" is a thing that works against the crew. Thus, most centcom roles are not antags, with the exception of Deathsquid, and even then some DS teams actually RESCUE crewmembers (weird, right?)

Anyway, since there is lots of murder, cruel medical experimentation, slavery and total lack of ethics on SS13, sometimes all at once, I wouldn't be surprised if we actually were the bad guys. At least by comparison. Kind of like some of the 40K factions - evil by our standards, but meh by theirs.

Shadowlight213 commented:
That's sort of the point behind it. You better be careful about stealing those yellow gloves, while the archangel is around or they might just holy judgement your ass.

@octareenroon91 give me sprites, and I can. The wings are just as easy to add onto as the ears/tail. The hard part is adding the new bodypart itself. Adding new options is extremely simple codewise.

Shadowlight213 committed:
More WIP
Initial work on spacemove in atmosphere
Getting stunned now launches you and doubles the stun.

KorPhaeron commented:
>assuring us it wont be roundstart while arguing it wont be a balance issue for it to be roundstart

You can't fool me

Shadowlight213 commented:
@KorPhaeron angel wings specifically are NEVER going to be round start. Ever. No maintainer in thwir right mind would allow it. Im not even going to make ot a config. They are far too effective.
There is no balance issue for non angel wings at all though, because as I've said, the flying is angel only. All other wings would be cosmetic with a downside of a larger hitbox.
Although hitbox depends on the actual wing sprite.

PKPenguin321 commented:
Just make a wings jetpack for your antag instead of enabling bird fetishists please

Shadowlight213 commented:
I can comment out the wings from roundstart selection completely, rather than make it a config if it's triggering people too much.
@pkpenguin321 making a jetpack would be relatively simple, true. However this is partially a test on my half to see what I can accomplish with the mutant parts system.

PKPenguin321 commented:
suuuure it is

Shadowlight213 committed:
Change to angel species.

Shadowlight213 committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into Archangel

Shadowlight213 committed:
More WIP with changing to species.

Shadowlight213 committed:
Fixes merge conflicts

KorPhaeron commented:
@Shadowlight213 

So does this stuff work?

KorPhaeron commented:
Also I feel like this should probably take up your back slot somehow. Maybe add the species flag slot blocking stuff or something

Shadowlight213 commented:
@KorPhaeron yes, this actually does work.
I've tested the stun effect and the jetpack effect and they both work.
I still need to fix up the float effect as update gravity is  causing the float to sort of stutter as it tries to switch it off while the species code is trying to turn it on.

Jazaen commented:
Make wings a bodypart, and add wing transplantation surgery! BECAUSE KILLING SERVANTS OF HIGHER POWERS AND GRAFTING PARTS OF THEIR BODiES SEEMS LIKE SOMETHING NT WOULD DO.

Shadowlight213 commented:
@Jazaen No. because doing that would require me to literally throw out 80% of my work and start over from scratch.

Shadowlight213 committed:
Speed in nograv works correctly now
Dismemberment support added as well as a way to use the limbs of another species for dismemberment for visually identical species
Flying var changed to species flag

Shadowlight213 committed:
Merge conflict fix

Shadowlight213 commented:
Ran a testmerge round.
TODO reminders list
Add wings giving protection from chasms (laval?), wet floors
Make the no backpack restriction be removed if wearing a hardsuit
Add a downside/reason to not fly nonstop

Shadowlight213 committed:
NEW SPRITES!
make flying give immunity to chasms

Supermichael777 commented:
stamina damage

Shadowlight213 commented:
Stamina damage might work. Probably stamina damage gained by moving, with damage healing slowed while hovering.

KorPhaeron commented:
Is a cooldown not possible? Sprinting+self knockout was terrible

Shadowlight213 commented:
@KorPhaeron the stamina damage would be slower than sprinting.
Also I can make the flying deactivate shortly before knockout level.

KorPhaeron commented:
Finish this I wanna fly around

Shadowlight213 committed:
Conflict fixes

Shadowlight213 commented:
Ok, I decided not to include the stamina damage. This should be ready to merge i guess.

optimumtact commented:
@Shadowlight213 has this had a test merge?

Can you arrange for one for 24 hours or something?

KorPhaeron commented:
Not a recent test merge no

Shadowlight213 commented:
yeah. sure

KorPhaeron commented:
Is there any reason why an addwings button like purrbation wouldnt work?

Shadowlight213 commented:
@KorPhaeron It being a species allows me to have all the flying code localized to the species datum.

Shadowlight213 commented:
Also I think I broke cat ears somewhow. I need to look into this

Shadowlight213 committed:
Fixes cat ears

Shadowlight213 committed:
makes toggle purrbation work on angels

Shadowlight213 committed:
Fixes up layers so it looks better.
Removes restrain check on wings.

optimumtact commented:
Fix the conflicts and I'm fairly happy to pull the trigger on this

Shadowlight213 committed:
Fixes merge conflicts

Shadowlight213 commented:
@optimumtact done

phil235 commented on code/__DEFINES/misc.dm:
```diff
@@ -23,24 +23,25 @@
+#define BODY_WING_LAYER			13		//Optimum layer for wing mutant parts. below hair and headgear, and above shit on the body. Can't do better without directional layers
+#define BACK_LAYER				12
+#define EARS_LAYER				11
+#define GLASSES_LAYER			10
```
why the changes here? I mean outside of adding the wing layer.

Shadowlight213 replied:
Most of the other changes were done to make it look nicer. It's hard to explain it just codewise, but I moved all the stuff around until it looked more logical in game.

Cheridan commented:
the slope has been lubed

HankMoodyExperience commented:
>NOT GIVING THE LAWYER WINGS ROUNDSTART AND RENAMING HIM HARVEY BIRDMAN: ATTORNEY AT LAW

REFERENCE MISSED

Incoming5643 commented on code/modules/mob/living/carbon/human/species_types.dm:
```diff
@@ -725,3 +725,148 @@ SYNDICATE BLACK OPS
+	name = "Toggle Flight"
+	check_flags = AB_CHECK_CONSCIOUS|AB_CHECK_STUNNED
+	button_icon_state = "slimesplit"
+	background_icon_state = "bg_alien"
```
Quack

Incoming5643 commented on code/__DEFINES/misc.dm:
```diff
@@ -23,24 +23,25 @@
+#define SUIT_LAYER				16
+#define BELT_LAYER				15		//Possible make this an overlay of somethign required to wear a belt?
+#define SUIT_STORE_LAYER		14
+#define BODY_WING_LAYER			13		//Optimum layer for wing mutant parts. below hair and headgear, and above shit on the body. Can't do better without directional layers
```
BODY_BEHIND_LAYER
BODY_ADJ_LAYER
FRONT_MUTATIONS_LAYER

ARE THE DIRECTIONAL LAYERS

Shadowlight213 replied:
I tried that. Uniform was rendered above the wings when facing north. Honk honk.
